url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/mult/CheckMatrixVectorMultShape.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.dense.mult;
+
+import org.ejml.data.DenseMatrix64F;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+
+/**
+ * Checks to see if the input to a matrix vector mutiply is accepted or rejected correctly depending
+ * on the shape in the input matrices.  Java reflections is used to grab all functions
+ * with "mult" in its name and then it determins if any of matrices are transposed.
+ *
+ * @author Peter Abeles
+ */
+public class CheckMatrixVectorMultShape {
+
+    Class theClass;
+
+    public CheckMatrixVectorMultShape( Class theClass ) {
+        this.theClass = theClass;
+    }
+
+    /**
+     * Perform all shape input checks.
+     *
+     * @throws Throwable
+     */
+    public void checkAll()
+    {
+        int numChecked = 0;
+        Method methods[] = theClass.getMethods();
+
+        for( Method method : methods ) {
+            String name = method.getName();
+
+            // only look at function which perform matrix multiplcation
+            if( !name.contains("mult"))
+                continue;
+
+            boolean transA = false;
+
+            if( name.contains("TransA")) {
+                transA = true;
+            }
+
+            try {
+                checkPositive(method, transA);
+                checkNegative(method,transA);
+            } catch( Throwable e ) {
+                System.out.println("Failed on function: "+name);
+                e.printStackTrace();
+                fail("An exception was thrown");
+            }
+            numChecked++;
+        }
+
+        // make sure some functions were checked!
+        assertTrue(numChecked!=0);
+    }
+
+    /**
+     * See if the function can be called with matrices of the correct size
+     */
+    private void checkPositive(Method func, boolean transA) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+        DenseMatrix64F A,B;
+        DenseMatrix64F C = new DenseMatrix64F(2,1);
+
+        if( transA ) {
+            A = new DenseMatrix64F(4,2);
+        } else {
+            A = new DenseMatrix64F(2,4);
+        }
+
+        // should work for B as a column or row vector
+        B = new DenseMatrix64F(4,1);
+        func.invoke(null,A,B,C);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)